pkg/loadbalancer/flow/flow.go

```diff
@@ -17,50 +17,138 @@ limitations under the License.
 	"github.com/sirupsen/logrus"
 )
 
+// Flow holds flow data
 type Flow struct {
 	*nspAPI.Flow
 	nfqueueLoadBalancer types.NFQueueLoadBalancer
+	nftHandler types.NftHandler
 }
 
-func New(flow *nspAPI.Flow, options ...Option) (types.Flow, error) {
-	opts := &flowOptions{}
-	for _, opt := range options {
-		opt(opts)
+// New creates a new flow
+func New(flow *nspAPI.Flow, lb types.NFQueueLoadBalancer, nfth types.NftHandler) (types.Flow, error) {
+	if flow == nil {
+		return nil, fmt.Errorf("Flow:New: Create nil Flow")
+	}
+	logrus.Infof("Flow: New %s", flow.Name)
+	if lb == nil {
+		return nil, fmt.Errorf("Flow(%s):New: No NFQueueLoadBalancer", flow.Name)
+	}
+	if nfth == nil {
+		return nil, fmt.Errorf("Flow(%s):New: No NftHandler", flow.Name)
+	}
+
+	if flow.LocalPort != 0 {
+		dport, err := ensureSingleDport(flow)
+		if err != nil {
+			return nil, err
+		}
+		logrus.Infof("Flow:New: Port NAT: %v -> %v", dport, flow.LocalPort)
+
+		err = nfth.PortNATCreateSets(flow)
+		if err != nil {
+			return nil, err
+		}
+		err = nfth.PortNATSetAddresses(flow)
+		if err != nil {
+			nfth.PortNATDeleteSets(flow)
+			return nil, err
+		}
+
+		err = nfth.PortNATSet(
+			flow.Name, flow.Protocols, dport, uint(flow.LocalPort))
+		if err != nil {
+			nfth.PortNATDeleteSets(flow)
+			return nil, err
+		}
 	}
 
 	f := &Flow{
 		Flow:                flow,
-		nfqueueLoadBalancer: opts.nfqueueLoadBalancer,
+		nfqueueLoadBalancer: lb,
+		nftHandler:          nfth,
 	}
-
-	var err error
-	if f.nfqueueLoadBalancer != nil {
-		err = f.nfqueueLoadBalancer.SetFlow(flow)
+	if err := f.nfqueueLoadBalancer.SetFlow(f.Flow); err != nil {
+		if flow.LocalPort != 0 {
+			nfth.PortNATDelete(flow.Name)
+			nfth.PortNATDeleteSets(flow)
+		}
+		return nil, err
 	}
-	logrus.Infof("New flow: %v - err: %v", flow, err)
 	return f, nil
 }
 
+// Update updates a flow
 func (f *Flow) Update(flow *nspAPI.Flow) error {
-	if f.Flow == nil || !f.Flow.DeepEquals(flow) {
-		f.Flow = flow
-		logrus.Infof("Update flow: %v", f.Flow)
-		if f.nfqueueLoadBalancer != nil {
-			return f.nfqueueLoadBalancer.SetFlow(f.Flow)
+	logrus.Infof("Flow:Update: %v -> %v", f.Flow, flow)
+	if f.Flow.DeepEquals(flow) {
+		// Not changed
+		logrus.Debugf("Flow:Update: to same")
+		return nil
+	}
+	if f.Flow.Name != flow.Name {
+		logrus.Warningf("Flow:Update: name %v -> %v", f.Flow.Name, flow.Name)
+		return fmt.Errorf("Flow:Update Name is not allowed")
+	}
+
+	// Check port-NAT
+	if flow.LocalPort != 0 {
+		dport, err := ensureSingleDport(flow)
+		if err != nil {
+			return err
+		}
+		oldDport, _ := ensureSingleDport(f.Flow)
+
+		if flow.LocalPort != f.Flow.LocalPort || dport != oldDport {
+			logrus.Infof("Flow:Update: Port NAT: %v -> %v", dport, flow.LocalPort)
+			err = f.nftHandler.PortNATSet(
+				flow.Name, flow.Protocols, dport, uint(flow.LocalPort))
+			if err != nil {
+				return err
+			}
+		}
+		err = f.nftHandler.PortNATSetAddresses(flow)
+		if err != nil {
+			return err
```

Comment:
is it worth to clean-up here?
f.nftHandler.PortNATDelete(f.Flow.Name)

Reply:
Good comment. But the more I think about it the less certain I become about the best action.

The situation is that the flow is updated either with an incorrect configuration or there is some problem with nft. What we don't want is to create a configuration that doesn't allow the user to correct things.

Reply:
I decided to *not* cleanup anything.

The localPort is != 0 and a natural recovery user action would be to delete the flow and re-create it. On delete everything will be cleaned up.